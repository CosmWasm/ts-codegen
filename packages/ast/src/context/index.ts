export * from './context';
export * from './imports';