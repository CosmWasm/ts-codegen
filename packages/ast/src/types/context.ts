import { IDLObject, JSONSchema } from "@cosmwasm/ts-codegen-types";
import { MessageBuilderOptions, MessageComposerOptions, ReactQueryOptions, RecoilOptions, TSClientOptions, TSTypesOptions, UseContractsOptions } from "./plugins";

export interface ContractInfo {
  schemas: JSONSchema[];
  responses?: Record<string, JSONSchema>;
  idlObject?: Partial<IDLObject>;
};
export interface RenderOptions {
  enabled?: boolean;
  types?: TSTypesOptions;
  recoil?: RecoilOptions;
  messageComposer?: MessageComposerOptions;
  messageBuilder?: MessageBuilderOptions;
  client?: TSClientOptions;
  reactQuery?: ReactQueryOptions;
  useContractsHook?: UseContractsOptions;
}

export interface ProviderInfo {
  classname: string,
  filename: string,
  basename: string,
}
