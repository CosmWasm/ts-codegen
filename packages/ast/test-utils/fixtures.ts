// @ts-nocheck
import { JSONSchema } from '../src/types';

import canExecuteRelayResponseData from '../../../__fixtures__/vectis/govec/can_execute_relay_response.json';
import cosmosMsgForEmptyData from '../../../__fixtures__/vectis/govec/cosmos_msg_for__empty.json';
import executeMsgForEmptyData from '../../../__fixtures__/vectis/govec/execute_msg_for__empty.json';
import infoResponseData from '../../../__fixtures__/vectis/govec/info_response.json';
import relayTransactionData from '../../../__fixtures__/vectis/govec/relay_transaction.json';

import nftIdlVersionData from '../../../__fixtures__/idl-version/accounts-nft/account-nft.json';

export const nftIdlVersion: JSONSchema = nftIdlVersionData;
export const vectisCanExecuteRelayResponse: JSONSchema = canExecuteRelayResponseData;
export const vectisCosmosMsgForEmpty: JSONSchema = cosmosMsgForEmptyData;
export const vectisExecuteMsgForEmpty: JSONSchema = executeMsgForEmptyData;
export const vectisInfoResponse: JSONSchema = infoResponseData;
export const vectisRelayTransaction: JSONSchema = relayTransactionData;
