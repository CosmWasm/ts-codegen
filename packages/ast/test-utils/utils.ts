import generate from '@babel/generator';
import { RenderOptions } from '@cosmwasm/ts-codegen-ast/types';
import { ExecuteMsg, IDLObject, JSONSchema, QueryMsg } from '@cosmwasm/ts-codegen-types';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { join } from 'path';

import { RenderContext } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
};

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
};

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options);
};

interface GlobContractLegacy {
  name: `/${string}.json`;
  content: JSONSchema;
}

interface GlobContract {
  name: `/${string}.json`;
  content: IDLObject;
}

const fixtureDir = join(__dirname, '../../../__fixtures__');
const globCache: Record<string, GlobContractLegacy[] | GlobContract[]> = {};

export const globIdlBasedContracts = (p: string): GlobContract[] => {
  if (globCache[p]) return globCache[p] as GlobContract[];
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const globLegacyContracts = (p: string): GlobContractLegacy[] => {
  if (globCache[p]) return globCache[p] as GlobContractLegacy[];
  // @ts-ignore
  const contracts: GlobContractLegacy[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const getMsgExecuteLegacyFixture = (
  p: string,
  name: `/${string}.json`
): ExecuteMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as ExecuteMsg;
};

export const getMsgQueryLegacyFixture = (
  p: string,
  name: `/${string}.json`
): QueryMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as QueryMsg;
};

export const getLegacyFixture = (
  p: string,
  name: `/${string}.json`
): JSONSchema => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as JSONSchema;
};
