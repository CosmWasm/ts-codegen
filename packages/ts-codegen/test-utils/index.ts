import {
  BuilderContext,
  ContractInfo,
  ReactQueryOptions,
  RenderOptions,
  TSTypesOptions,
} from "wasm-ast-types";
import { ReactQueryPlugin } from "../src/plugins/react-query";
import { TypesPlugin } from "../src/plugins/types";
import { ClientPlugin } from "../src/plugins/client";
import { MessageComposerPlugin } from "../src/plugins/message-composer";
import { RecoilPlugin } from "../src/plugins/recoil";
import { MessageBuilderPlugin } from "../src/plugins/message-builder";
import { ContractsContextProviderPlugin } from "../src/plugins/provider";
import { ContractsProviderBundlePlugin } from "../src/plugins/provider-bundle";
import { TSBuilderOptions } from "../src/builder";
import { createDefaultContractInfo } from "../src/utils/contracts";
import {
  TestContractsContextProviderPlugin,
  TestClientPlugin,
  TestMessageComposerPlugin,
  TestContractsProviderBundlePlugin,
} from "./testPlugins";
import deepmerge from "deepmerge";

export const testDefaultOptions: RenderOptions = {
  enabled: true,
  types: {
    enabled: true,
    itemsUseTuples: false,
    aliasExecuteMsg: false,
  },
  client: {
    enabled: true,
    execExtendsQuery: true,
    noImplicitOverride: false,
  },
  recoil: {
    enabled: true,
  },
  messageComposer: {
    enabled: true,
  },
  messageBuilder: {
    enabled: true,
  },
  reactQuery: {
    enabled: true,
    optionalClient: false,
    version: "v4",
    mutations: false,
    camelize: true,
    queryKeys: false,
  },
  useContractsHook: {
    enabled: true,
  },
};

export async function generateReactQuery(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: ReactQueryOptions
) {
  const options = opts
    ? deepmerge(testDefaultOptions, { reactQuery: opts })
    : testDefaultOptions;

  await new ReactQueryPlugin(options).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateTypes(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: TSTypesOptions
) {
  await new TypesPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateClient(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestClientPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new ClientPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageComposer(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestMessageComposerPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new MessageComposerPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageBuilder(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new MessageBuilderPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateRecoil(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new RecoilPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateContractHooks(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext
) {
  await new TestContractsContextProviderPlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, contractName, contractInfo);

  await new TestContractsProviderBundlePlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, undefined, contractInfo ?? createDefaultContractInfo());
}
