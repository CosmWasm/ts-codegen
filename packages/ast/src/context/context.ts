import * as t from '@babel/types';
import { JSONSchema } from "@cosmwasm/ts-codegen-types";
import { refLookup } from "../utils";
import { convertUtilsToImportList, getImportStatements, UtilMapping } from "./imports";
import deepmerge from "deepmerge";
import { basename, extname } from 'path'
import { ContractInfo, ProviderInfo, RenderOptions } from '../types';

export interface IContext {
  refLookup($ref: string): JSONSchema;
  addUtil(util: string): void;
  getImports(registeredUtils?: UtilMapping, filepath?: string): (t.ImportNamespaceSpecifier | t.ImportDeclaration | t.ImportDefaultSpecifier)[];
}

export interface IRenderContext<TOpt = RenderOptions> extends IContext {
  contract: ContractInfo;
  options: TOpt;

  addProviderInfo(contractName: string, type: string, classname: string, filename: string): void;
  getProviderInfos(): {
    [key: string]: {
      [key: string]: ProviderInfo;
    };
  };
}

export const defaultOptions: RenderOptions = {
  enabled: true,
  types: {
    enabled: true,
    itemsUseTuples: false,
    aliasExecuteMsg: false
  },
  client: {
    enabled: true,
    execExtendsQuery: true,
    noImplicitOverride: false,
  },
  recoil: {
    enabled: false
  },
  messageComposer: {
    enabled: false
  },
  messageBuilder: {
    enabled: false,
  },
  reactQuery: {
    enabled: false,
    optionalClient: false,
    version: 'v4',
    mutations: false,
    camelize: true,
    queryKeys: false
  },
  useContractsHook: {
    enabled: false
  }
};

export const getDefinitionSchema = (schemas: JSONSchema[]): JSONSchema => {
  const aggregateSchema = {
    definitions: {
      //
    }
  };

  schemas.forEach(schema => {
    schema.definitions = schema.definitions || {};
    aggregateSchema.definitions = {
      ...aggregateSchema.definitions,
      ...schema.definitions
    };
  });

  return aggregateSchema;
};

export class BuilderContext {
  providers: {
    [key: string]: {
      [key: string]: ProviderInfo;
    };
  } = {};

  addProviderInfo(contractName: string, type: string, classname: string, filename: string): void {
    if (!this.providers[contractName]) {
      this.providers[contractName] = {}
    }

    this.providers[contractName][type] = {
      classname,
      filename,
      basename: basename(filename, extname(filename))
    };
  }
  getProviderInfos(): {
    [key: string]: {
      [key: string]: ProviderInfo;
    };
  } {
    return this.providers;
  }
}

/**
 * context object for generating code.
 * only mergeDefaultOpt needs to implementing for combine options and default options.
 * @param TOpt option type
 */
export abstract class RenderContextBase<TOpt = RenderOptions> implements IRenderContext<TOpt> {
  builderContext: BuilderContext;
  contract: ContractInfo;
  utils: Record<string, boolean> = {};
  schema: JSONSchema;
  options: TOpt;

  constructor(
    contract: ContractInfo,
    options?: TOpt,
    builderContext?: BuilderContext
  ) {
    this.contract = contract;
    this.schema = getDefinitionSchema(contract.schemas);
    this.options = this.mergeDefaultOpt(options);
    this.builderContext = builderContext ?? new BuilderContext();
  }
  /**
   * merge options and default options
   * @param options
   */
  abstract mergeDefaultOpt(options: TOpt): TOpt;
  refLookup($ref: string): JSONSchema {
    return refLookup($ref, this.schema)
  }
  addUtil(util: string): void {
    this.utils[util] = true;
  }
  addProviderInfo(contractName: string, type: string, classname: string, filename: string): void {
    this.builderContext.addProviderInfo(contractName, type, classname, filename);
  }
  getProviderInfos(): {
    [key: string]: {
      [key: string]: ProviderInfo;
    };
  } {
    return this.builderContext.providers;
  }
  getImports(registeredUtils?: UtilMapping, filepath?: string): (t.ImportNamespaceSpecifier | t.ImportDeclaration | t.ImportDefaultSpecifier)[] {
    return getImportStatements(
      convertUtilsToImportList(
        this,
        Object.keys(this.utils),
        registeredUtils,
      ),
      filepath
    );
  }
}

export class RenderContext extends RenderContextBase {
  mergeDefaultOpt(options: RenderOptions): RenderOptions {
    return deepmerge(defaultOptions, options ?? {});
  }
}
