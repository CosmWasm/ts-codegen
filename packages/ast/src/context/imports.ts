import * as t from '@babel/types';
import { dirname, extname,relative } from 'path';

import { importAs, importStmt } from '../utils';
import { RenderContext } from './context';


export interface ImportObj {
  type: 'import' | 'default' | 'namespace';
  name: string;
  path: string;
  importAs?: string;
}

export type GetUtilFn = (<TContext = RenderContext>(...args: any[]) => (context: TContext) => ImportObj);
export type UtilMapping = {
  [key: string]:
  | ImportObj
  | string
  | GetUtilFn
};

const makeReactQuerySwitch = (varName: string) => {
  return (context: RenderContext) => {
    switch (context.options.reactQuery.version) {
    case 'v4':
      return {
        type: 'import',
        path: '@tanstack/react-query',
        name: varName
      };
    case 'v3':
    default:
      return {
        type: 'import',
        path: 'react-query',
        name: varName
      };
    }
  };
};

export const UTILS = {
  selectorFamily: 'recoil',
  MsgExecuteContract: 'cosmjs-types/cosmwasm/wasm/v1/tx',
  MsgExecuteContractEncodeObject: '@cosmjs/cosmwasm-stargate',
  Coin: '@cosmjs/amino',
  toUtf8: '@cosmjs/encoding',
  StdFee: '@cosmjs/amino',
  CosmWasmClient: '@cosmjs/cosmwasm-stargate',
  ExecuteResult: '@cosmjs/cosmwasm-stargate',
  SigningCosmWasmClient: '@cosmjs/cosmwasm-stargate',

  // react-query
  useQuery: makeReactQuerySwitch('useQuery'),
  UseQueryOptions: makeReactQuerySwitch('UseQueryOptions'),
  useMutation: makeReactQuerySwitch('useMutation'),
  UseMutationOptions: makeReactQuerySwitch('UseMutationOptions')

};

export const UTIL_HELPERS = [
  '__contractContextBase__',
];

export const convertUtilsToImportList = (
  context: RenderContext,
  utils: string[],
  registeredUtils?: UtilMapping
): ImportObj[] => {
  return utils.map((util) => {
    let result = null;

    if (registeredUtils) {
      result = convertUtil(context, util, registeredUtils);

      if (result) {
        return result;
      }
    }

    result = convertUtil(context, util, UTILS);

    if (result) {
      return result;
    }

    throw new Error(`missing Util! ::[${util}]`);
  });
};

export const convertUtil = (
  context: RenderContext,
  util: string,
  registeredUtils: any
): ImportObj => {
  if (!registeredUtils.hasOwnProperty(util)) return null;
  if (typeof registeredUtils[util] === 'string') {
    return {
      type: 'import',
      path: registeredUtils[util],
      name: util
    };
  } else if (typeof registeredUtils[util] === 'function') {
    return registeredUtils[util](context);
  } else {
    return registeredUtils[util];
  }
};


// __helpers__
export const getImportStatements = (
  list: ImportObj[],
  filepath?: string
): (t.ImportNamespaceSpecifier | t.ImportDeclaration | t.ImportDefaultSpecifier)[] => {

  // swap helpers with helpers file...
  const modifiedImports = list.map(imp => {
    if (filepath && UTIL_HELPERS.includes(imp.path)) {
      const name = imp.path.replace(/__/g, '');
      return {
        ...imp,
        path: getRelativePath(filepath, `./${name}`)
      };
    }
    return imp;
  });

  const imports = modifiedImports.reduce((m: Record<string, ImportObj[]>, obj) => {
    m[obj.path] = m[obj.path] || [];
    const exists = m[obj.path].find((el: ImportObj) =>
      el.type === obj.type && el.path === obj.path && el.name === obj.name);

    // MARKED AS NOT DRY [google.protobuf names]
    // TODO some have google.protobuf.Any shows up... figure out the better way to handle this
    if (/\./.test(obj.name)) {
      obj.name = obj.name.split('.')[obj.name.split('.').length - 1];
    }

    if (!exists) {
      m[obj.path].push(obj);
    }
    return m;
  }, {});


  return Object.entries(imports)
    .reduce((m, [importPath, imports]: [string, ImportObj[]]) => {
      const defaultImports = imports.filter(a => a.type === 'default');
      if (defaultImports.length) {
        if (defaultImports.length > 1) throw new Error('more than one default name NOT allowed.');
        m.push(
          t.importDeclaration(
            [
              t.importDefaultSpecifier(
                t.identifier(defaultImports[0].name)
              )
            ],
            t.stringLiteral(defaultImports[0].path)
          )
        );
      }
      const namedImports = imports.filter(a => a.type === 'import' && (!a.importAs || (a.name === a.importAs)));
      if (namedImports.length) {
        m.push(importStmt(namedImports.map(i => i.name), namedImports[0].path));
      }
      const aliasNamedImports = imports.filter(a => a.type === 'import' && (a.importAs && (a.name !== a.importAs)));
      aliasNamedImports.forEach(imp => {
        m.push(importAs(imp.name, imp.importAs, imp.path));
      });

      const namespaced = imports.filter(a => a.type === 'namespace');
      if (namespaced.length) {
        if (namespaced.length > 1) throw new Error('more than one namespaced name NOT allowed.');
        m.push(
          t.importDeclaration(
            [
              t.importNamespaceSpecifier(
                t.identifier(namespaced[0].name)
              )
            ],
            t.stringLiteral(namespaced[0].path)
          )
        );
      }
      return m;
    }, []);
};

export const getRelativePath = (f1: string, f2: string) => {
  const rel = relative(dirname(f1), f2);
  let importPath = rel.replace(extname(rel), '');
  if (!/^\./.test(importPath)) importPath = `./${importPath}`;
  return importPath;
};