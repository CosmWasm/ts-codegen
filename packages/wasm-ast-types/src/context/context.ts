import { JSONSchema } from "../types";
import { refLookup } from "../utils";
import { convertUtilsToImportList, getImportStatements, UtilMapping } from "./imports";
import deepmerge from "deepmerge";

/// Plugin Types
export interface ReactQueryOptions {
    enabled?: boolean;
    optionalClient?: boolean;
    version?: 'v3' | 'v4';
    mutations?: boolean;
    camelize?: boolean;
    queryKeys?: boolean
    queryFactory?: boolean
}

export interface AbstractAppOptions {
  enabled?: boolean;
  clientPrefix?: string;
}

export interface TSClientOptions {
    enabled?: boolean;
    execExtendsQuery?: boolean;
    noImplicitOverride?: boolean;
}
export interface MessageComposerOptions {
    enabled?: boolean;
}
export interface MsgBuilderOptions {
    enabled?: boolean;
}
export interface RecoilOptions {
    enabled?: boolean;
}
export interface TSTypesOptions {
    enabled?: boolean;
    // deprecated
    aliasExecuteMsg?: boolean;
    aliasEntryPoints?: boolean;
}

/// END Plugin Types

interface KeyedSchema {
    [key: string]: JSONSchema;
}
export interface IDLObject {
    contract_name: string;
    contract_version: string;
    idl_version: string;
    instantiate: JSONSchema;
    execute: JSONSchema;
    query: JSONSchema;
    migrate: JSONSchema;
    sudo: JSONSchema;
    responses: KeyedSchema;
}

export interface ContractInfo {
    schemas: JSONSchema[];
    responses?: Record<string, JSONSchema>;
    idlObject?: IDLObject;
};
export interface RenderOptions {
    enabled?: boolean;
    types?: TSTypesOptions;
    recoil?: RecoilOptions;
    messageComposer?: MessageComposerOptions;
    msgBuilder?: MsgBuilderOptions;
    client?: TSClientOptions;
    reactQuery?: ReactQueryOptions;
    abstractApp?: AbstractAppOptions
}


export interface IContext {
  refLookup($ref: string);
  addUtil(util: string);
  addUtils(util: string[]);
  getImports(registeredUtils?: UtilMapping);
}

export interface IRenderContext<TOpt = RenderOptions> extends IContext {
    contract: ContractInfo;
    options: TOpt;
}

export const defaultOptions: RenderOptions = {
    enabled: true,
    types: {
        enabled: true,
        aliasExecuteMsg: false
    },
    client: {
        enabled: true,
        execExtendsQuery: true,
        noImplicitOverride: false,
    },
    recoil: {
        enabled: false
    },
    messageComposer: {
        enabled: false
    },
    msgBuilder: {
      enabled: false,
    },
    reactQuery: {
        enabled: false,
        optionalClient: false,
        version: 'v3',
        mutations: false,
        camelize: true,
        queryKeys: false
    },
    abstractApp: {
        enabled: false,
        clientPrefix: 'App'
    }
};

export const getDefinitionSchema = (schemas: JSONSchema[]): JSONSchema => {
    const aggregateSchema = {
        definitions: {
            //
        }
    };

    schemas.forEach(schema => {
        schema.definitions = schema.definitions || {};
        aggregateSchema.definitions = {
            ...aggregateSchema.definitions,
            ...schema.definitions
        };
    });

    return aggregateSchema;
};

/**
 * context object for generating code.
 * only mergeDefaultOpt needs to implementing for combine options and default options.
 * @param TOpt option type
 */
export abstract class RenderContextBase<TOpt = RenderOptions> implements IRenderContext<TOpt> {
    contract: ContractInfo;
    utils: string[] = [];
    schema: JSONSchema;
    options: TOpt;
    constructor(
        contract: ContractInfo,
        options?: TOpt
    ) {
        this.contract = contract;
        this.schema = getDefinitionSchema(contract.schemas);
        this.options = this.mergeDefaultOpt(options);
    }
    /**
     * merge options and default options
     * @param options
     */
    abstract mergeDefaultOpt(options: TOpt): TOpt;
    refLookup($ref: string) {
      return refLookup($ref, this.schema)
    }
    addUtil(util: string) {
        this.utils[util] = true;
    }
    addUtils(utils: string[]) {
        utils.forEach(util => {
            this.utils[util] = true;
        });
    }
    getImports(registeredUtils?: UtilMapping) {
        return getImportStatements(
            convertUtilsToImportList(
                this,
                Object.keys(this.utils),
                registeredUtils,
            )
        );
    }
}

export class RenderContext extends RenderContextBase{
  mergeDefaultOpt(options: RenderOptions): RenderOptions {
    return deepmerge(defaultOptions, options ?? {});
  }
}
