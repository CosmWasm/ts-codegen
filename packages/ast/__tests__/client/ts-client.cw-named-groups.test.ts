import execute_msg from '../../../../../__fixtures__/daodao/cw-named-groups/execute_msg.json';

import {
    createQueryClass,
    createExecuteClass,
    createExecuteInterface,
    createTypeInterface
} from '../client'
import { RenderContext } from '../../context';
import { expectCode, makeContext } from '../../../test-utils';

const ctx = makeContext(execute_msg);

it('execute_msg', () => {
    expectCode(createTypeInterface(
        ctx,
        execute_msg
    ))
})


it('query classes', () => {
    expectCode(createQueryClass(
        ctx,
        'SG721QueryClient',
        'SG721ReadOnlyInstance',
        execute_msg
    ))
});

it('execute classes array types', () => {
    expectCode(createExecuteClass(
        ctx,
        'SG721Client',
        'SG721Instance',
        null,
        execute_msg
    ))
});

it('execute interfaces no extends', () => {
    expectCode(createExecuteInterface(
        ctx,
        'SG721Instance',
        null,
        execute_msg
    ))
});
