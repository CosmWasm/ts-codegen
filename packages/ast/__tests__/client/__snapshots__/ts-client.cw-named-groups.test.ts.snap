// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute classes array types 1`] = `
"export class SG721Client implements SG721Instance {
  client: SigningCosmWasmClient;
  sender: string;
  contractAddress: string;
  constructor(client: SigningCosmWasmClient, sender: string, contractAddress: string) {
    this.client = client;
    this.sender = sender;
    this.contractAddress = contractAddress;
    this.update = this.update.bind(this);
    this.removeGroup = this.removeGroup.bind(this);
    this.updateOwner = this.updateOwner.bind(this);
  }
  update = async ({
    addressesToAdd,
    addressesToRemove,
    group
  }: {
    addressesToAdd?: string[];
    addressesToRemove?: string[];
    group: string;
  }, fee_: number | StdFee | "auto" = "auto", memo?: string, funds_?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      update: {
        addresses_to_add: addressesToAdd,
        addresses_to_remove: addressesToRemove,
        group
      }
    }, fee_, memo_, funds_);
  };
  removeGroup = async ({
    group
  }: {
    group: string;
  }, fee_: number | StdFee | "auto" = "auto", memo?: string, funds_?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      remove_group: {
        group
      }
    }, fee_, memo_, funds_);
  };
  updateOwner = async ({
    owner
  }: {
    owner: string;
  }, fee_: number | StdFee | "auto" = "auto", memo?: string, funds_?: Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      update_owner: {
        owner
      }
    }, fee_, memo_, funds_);
  };
}"
`;

exports[`execute interfaces no extends 1`] = `
"export interface SG721Instance {
  contractAddress: string;
  sender: string;
  update: ({
    addressesToAdd,
    addressesToRemove,
    group
  }: {
    addressesToAdd?: string[];
    addressesToRemove?: string[];
    group: string;
  }, fee_?: number | StdFee | "auto", memo_?: string, funds_?: Coin[]) => Promise<ExecuteResult>;
  removeGroup: ({
    group
  }: {
    group: string;
  }, fee_?: number | StdFee | "auto", memo_?: string, funds_?: Coin[]) => Promise<ExecuteResult>;
  updateOwner: ({
    owner
  }: {
    owner: string;
  }, fee_?: number | StdFee | "auto", memo_?: string, funds_?: Coin[]) => Promise<ExecuteResult>;
}"
`;

exports[`execute_msg 1`] = `"export type ExecuteMsg = ExecuteMsg;"`;

exports[`query classes 1`] = `
"export class SG721QueryClient implements SG721ReadOnlyInstance {
  client: CosmWasmClient;
  contractAddress: string;
  constructor(client: CosmWasmClient, contractAddress: string) {
    this.client = client;
    this.contractAddress = contractAddress;
    this.update = this.update.bind(this);
    this.removeGroup = this.removeGroup.bind(this);
    this.updateOwner = this.updateOwner.bind(this);
  }
  update = async ({
    addressesToAdd,
    addressesToRemove,
    group
  }: {
    addressesToAdd?: string[];
    addressesToRemove?: string[];
    group: string;
  }): Promise<UpdateResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      update: {
        addresses_to_add: addressesToAdd,
        addresses_to_remove: addressesToRemove,
        group
      }
    });
  };
  removeGroup = async ({
    group
  }: {
    group: string;
  }): Promise<RemoveGroupResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      remove_group: {
        group
      }
    });
  };
  updateOwner = async ({
    owner
  }: {
    owner: string;
  }): Promise<UpdateOwnerResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      update_owner: {
        owner
      }
    });
  };
}"
`;
